File: 算法训练营/04-栈队列/239/单调队列.go
/*
 * @lc app=leetcode.cn id=239 lang=golang
 *
 * [239] 滑动窗口最大值
 */

// @lc code=start
func maxSlidingWindow(nums []int, k int) []int {
	stack := make([]int, 0, k)
	// 准备数据
	for i, v := range nums[:k] {
		for len(stack) > 0 && nums[stack[len(stack)-1]] <= v {
			stack = stack[:len(stack)-1] //pop(-1)
		}
		stack = append(stack, i)
	}

	rets := make([]int, 0)
	rets = append(rets, nums[stack[0]])
	for i, v := range nums[k:] {
		i += k
		if i-k >= stack[0] {
			// pop(0)
			stack = stack[1:]
		}
		for len(stack)>0 && nums[stack[len(stack)-1]] <= v {
			// pop(-1)
			stack = stack[:len(stack)-1] //pop(-1)
		}
		stack = append(stack, i)
		rets = append(rets, nums[stack[0]])
	}
	return rets
}

// @lc code=end

File: LeetCode/0239/单调队列.py
#
# @lc app=leetcode.cn id=239 lang=python3
#
# [239] 滑动窗口最大值
#

# @lc code=start
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        single_queue = []
        
        for i in range(k-1):
            while single_queue:
                if nums[single_queue[-1]] < nums[i]:
                    single_queue.pop()
                else:
                    break
            single_queue.append(i)
        
        max_vars = []
        for i in range(k-1, len(nums)):
            
            while single_queue:
                if nums[single_queue[-1]] < nums[i]:
                    single_queue.pop()
                else:
                    break
            single_queue.append(i)
            
            max_vars.append(nums[single_queue[0]])

            if i - single_queue[0] >= k-1:
                single_queue = single_queue[1:]
            
        return max_vars

# @lc code=end